File: Heap/maxHeap.cpp
#include <iostream>
#include <vector>

class MaxHeap {
private:
    std::vector<int> heap;

    void heapify(int index) {
        int largest = index;
        int leftChild = 2 * index + 1;
        int rightChild = 2 * index + 2;

        // Compare with left child
        if (leftChild < heap.size() && heap[leftChild] > heap[largest])
            largest = leftChild;

        // Compare with right child
        if (rightChild < heap.size() && heap[rightChild] > heap[largest])
            largest = rightChild;

        // If the largest is not the current index, swap and recursively heapify
        if (largest != index) {
            std::swap(heap[index], heap[largest]);
            heapify(largest);
        }
    }

public:
    void insert(int value) {
        // Insert the new element at the end
        heap.push_back(value);

        // Fix the max heap property
        int index = heap.size() - 1;
        while (index > 0 && heap[(index - 1) / 2] < heap[index]) {
            std::swap(heap[index], heap[(index - 1) / 2]);
            index = (index - 1) / 2;
        }
    }

    int extractMax() {
        if (heap.empty()) {
            std::cout << "Heap underflow: Cannot extract from an empty heap." << std::endl;
            return -1; // Some sentinel value indicating an error
        }

        // The maximum element is at the root
        int max = heap[0];

        // Replace the root with the last element
        heap[0] = heap.back();
        heap.pop_back();

        // Fix the max heap property by calling heapify
        heapify(0);

        return max;
    }

    int findMax() const {
        if (heap.empty()) {
            std::cout << "Heap is empty: Cannot find maximum element." << std::endl;
            return -1; // Some sentinel value indicating an error
        }

        // The maximum element is at the root
        return heap[0];
    }

    int getHeapSize() const {
        return heap.size();
    }

    void printHeap() const {
        std::cout << "Max Heap: ";
        for (int value : heap) {
            std::cout << value << " ";
        }
        std::cout << std::endl;
    }
};

int main() {
    MaxHeap maxHeap;

    maxHeap.insert(10);
    maxHeap.insert(20);
    maxHeap.insert(15);
    maxHeap.insert(40);
    maxHeap.insert(50);
    maxHeap.insert(25);

    maxHeap.printHeap();

    std::cout << "Max element: " << maxHeap.findMax() << std::endl;

    int max = maxHeap.extractMax();
    std::cout << "Extracted max element: " << max << std::endl;

    maxHeap.printHeap();

    std::cout << "Heap size: " << maxHeap.getHeapSize() << std::endl;

    return 0;
}

File: Heap/maxHeap.py
class MaxHeap:
    def __init__(self):
        self.heap = []

    def insert(self, value):
        self.heap.append(value)
        self._heapify_up()

    def extract_max(self):
        if not self.heap:
            return None  # Heap is empty
        if len(self.heap) == 1:
            return self.heap.pop()

        max_value = self.heap[0]
        self.heap[0] = self.heap.pop()
        self._heapify_down()
        return max_value

    def _heapify_up(self):
        index = len(self.heap) - 1
        while index > 0:
            parent_index = (index - 1) // 2
            if self.heap[index] > self.heap[parent_index]:
                self.heap[index], self.heap[parent_index] = self.heap[parent_index], self.heap[index]
                index = parent_index
            else:
                break

    def _heapify_down(self):
        index = 0
        while True:
            left_child_index = 2 * index + 1
            right_child_index = 2 * index + 2
            largest = index

            if left_child_index < len(self.heap) and self.heap[left_child_index] > self.heap[largest]:
                largest = left_child_index

            if right_child_index < len(self.heap) and self.heap[right_child_index] > self.heap[largest]:
                largest = right_child_index

            if largest != index:
                self.heap[index], self.heap[largest] = self.heap[largest], self.heap[index]
                index = largest
            else:
                break

# Example usage:
max_heap = MaxHeap()
max_heap.insert(10)
max_heap.insert(20)
max_heap.insert(15)
max_heap.insert(40)
max_heap.insert(50)
max_heap.insert(25)

print("Max Heap:", max_heap.heap)

max_value = max_heap.extract_max()
print("Extracted max value:", max_value)
print("Max Heap after extraction:", max_heap.heap)
